### PR TITLE
[19기 김영훈] Refactor: 상품 상세보기 View 멘토 리뷰 반영

### PR DESCRIPTION
**Pull request by fcfargo**
## "http://0.0.0.0:8000/products/7" 통신 결과 
{
    "result": [
        {
            "allergy": [
                {
                    "allergy": "우유",
                    "allergy_id": 1,
                    "id": "1"
                },
                {
                    "allergy": "대두",
                    "allergy_id": 2,
                    "id": "2"
                },
                {
                    "allergy": "땅콩",
                    "allergy_id": 3,
                    "id": "3"
                }
            ],
            "comment": "유기농 식물성 원료로 더 깔끔하고 건강한 비바니 초콜릿",
            "delivery_type": "햇볕배송/택배배송",
            "description_image": "https://images.unsplash.com/photo-1584270354950-6e55b749ae63?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=1000&q=80",
            "discount_rate": 0.1,
            "id": "1",
            "instruction": "전 상품의 패키지가 리뉴얼되었습니다. 아래 상품 상세 정보 확인 부탁드립니다.",
            "name": "로이스초콜릿",
            "packing_type": "냉장/종이포장",
            "price": 20000,
            "product_id": 7,
            "related_products": [
                {
                    "id": "1",
                    "name": "필란드자일리톨",
                    "price": 7700,
                    "product_id": 10,
                    "rel_img": "https://images.unsplash.com/photo-1576644461179-ddd318c669e4?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=1000&q=80"
                },
                {
                    "id": "2",
                    "name": "로이스초콜릿",
                    "price": 20000,
                    "product_id": 7,
                    "rel_img": "https://images.unsplash.com/photo-1587271644048-2fbb187de8d8?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1000&q=80"
                },
                {
                    "id": "3",
                    "name": "초코칩쿠키",
                    "price": 7500,
                    "product_id": 6,
                    "rel_img": "https://images.unsplash.com/photo-1497051788611-2c64812349fa?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=1000&q=80"
                },
                {
                    "id": "4",
                    "name": "레몬사탕",
                    "price": 2500,
                    "product_id": 8,
                    "rel_img": "https://images.unsplash.com/photo-1603202902092-3eb7a76ecf5f?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1000&q=80"
                },
                {
                    "id": "5",
                    "name": "오렌지초콜릿",
                    "price": 4900,
                    "product_id": 9,
                    "rel_img": "https://images.unsplash.com/photo-1490324412536-e044eae4b09a?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=1000&q=80"
                },
                {
                    "id": "6",
                    "name": "화이트초코",
                    "price": 3200,
                    "product_id": 2,
                    "rel_img": "https://images.unsplash.com/photo-1548907040-4baa42d10919?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1000&q=80"
                },
                {
                    "id": "7",
                    "name": "머랭쿠키",
                    "price": 9900,
                    "product_id": 4,
                    "rel_img": "https://images.unsplash.com/photo-1607919565782-d49c91cb46cd?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=1000&q=80"
                },
                {
                    "id": "8",
                    "name": "과일캔디",
                    "price": 3000,
                    "product_id": 1,
                    "rel_img": "https://images.unsplash.com/photo-1601493700625-9185417898cf?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=1000&q=80"
                },
                {
                    "id": "9",
                    "name": "[롯데제과]ABC초코 65g",
                    "price": 2000,
                    "product_id": 11,
                    "rel_img": ""
                },
                {
                    "id": "10",
                    "name": "초코쿠키",
                    "price": 4500,
                    "product_id": 5,
                    "rel_img": "https://images.unsplash.com/photo-1499636136210-6f4ee915583e?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=1000&q=80"
                }
            ],
            "review": null,
            "sale_unit": "1개",
            "size_image": "https://i.esdrop.com/d/NlsEZZFqTI.jpg",
            "thumbnail_image": "https://images.unsplash.com/photo-1587271644048-2fbb187de8d8?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1000&q=80",
            "weight_g": 0.1
        }
    ]
}

## 수정 사항 간략한 한줄 요약
- 멘토 리뷰를 반영하여 상품 상세보기 View 코드를 refactoring하였습니다.
		 
## 수정 사항들 자세한 내용
-  product_id의 default값 추가
-  예외 추가
   - product_id값이 들어오지 않은 경우
   - 존재하지 않는 product_id값이 들어오는 경우
- 기존에 있던 변수 allergy_info 지우고, manytomany참조로 allergy_list값을 가져옴 
- product_id, allergy_id 값을 추가

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [X] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [X] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [X] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [X] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

### CODE REVIEW

File: products/views.py [hunk 1]
```diff
@@ -42,4 +44,58 @@ def get(self,request):
 
         } for product in products]
 
-        return JsonResponse({'results':results}, status=200)
\ No newline at end of file
+        return JsonResponse({'results':results}, status=200)
+
+class ProductDetailView(View):
+    def get(self, request, product_id):
+        product                 = Product.objects.get(id=product_id)
```
Comment (devops-andrew) [inline products/views.py]: product_id를 받는 부분 너무 착하네요. 제대로 전달안될경우의 익셉션이나, 안들어올경우의 핸들도 추가 부탁드립니다.

File: products/views.py [hunk 2]
```diff
@@ -42,4 +44,58 @@ def get(self,request):
...
+        product_info            = product.productinformation
+        allergy_info            = AllergyProduct.objects.filter(
+            product_information=product_info.id
+        )
+        
+        allergy_list            = [{
+            'id'      : '{}'.format(index+1),
```
Comment (devops-andrew) [inline products/views.py]: id 값을 알러지 본연의 id 값이 아닌 자료의 인덱스순서를 맞춰주려는 이유가 있나요?

File: products/views.py [hunk 3]
```diff
@@ -42,4 +44,58 @@ def get(self,request):
...
+        } for index in range(len(allergy_info))]
+
+        picked_related_products = self.pick_related_product([
+            related_product for related_product in Product.objects.all()
+        ])
+        
+        related_product_list    = self.make_related_product_list(picked_related_products) 
+#
+        result                  = [{
+            'id'                : "1",
```
Comment (devops-andrew) [inline products/views.py:85]: 이렇게 아이디를 고정하는 이유는요? 불러온 프로덕트 객체자체의 PK아이디를 줘야 하는게 아닌가요?

File: products/views.py [hunk 4]
```diff
@@ -42,4 +44,58 @@ def get(self,request):
...
+        product                 = Product.objects.get(id=product_id)
+        product_info            = product.productinformation
+        allergy_info            = AllergyProduct.objects.filter(
+            product_information=product_info.id
+        )
+        
+        allergy_list            = [{
+            'id'      : '{}'.format(index+1),
+            'allergy' : allergy_info[index].allergy.name
+        } for index in range(len(allergy_info))]
```
Comment (devops-andrew) [inline products/views.py]:
```suggestion        
        allergy_list            = [{
            'id'      : allergy.id,
            'allergy' : allergy.name
        } for allergy in product_info.allergy.all()]
```
중간테이블을 따로 색인할게 아니라. MtoM 적용했다면 그냥 이렇게 하는게 맞지 않나요?

**Review by devops-andrew — CHANGES_REQUESTED** [review]
확인해주세요.

File: products/views.py [hunk 5]
```diff
@@ -42,4 +45,81 @@ def get(self,request):
 
         } for product in products]
 
-        return JsonResponse({'results':results}, status=200)
\ No newline at end of file
+        return JsonResponse({'results':results}, status=200)
+
+class ProductDetailView(View):
+    def get(self, request, pk):
```
[the remark this answers is not in the document]
Comment (fcfargo) [reply products/views.py]: product_id의 디폴트 값을  None으로 설정하라는 부분 반영하겠습니다. 다만 한 가지 의아한 점은, 어제 피드백을 받은 후 매개변수 pk를 product_id로 수정하여 pull request를 했음에도, 코드에 반영되지 않았다는 점이네요... 바로 아래에서 지적해주신  urls.py의 product_id부분도 마찬가지입니다. 추후 다시 문제가 발생할 지도 몰라 말씀드립니다.

File: products/urls.py [hunk 6]
```diff
@@ -1,9 +1,9 @@
...
-from products.views import CategoryView,ProductView
+from products.views import CategoryView,ProductView,ProductDetailView
 
 urlpatterns = [
     path('/category', CategoryView.as_view()),
     path('/product', ProductView.as_view()),
-
-]
\ No newline at end of file
+    path('/<int:pk>', ProductDetailView.as_view()),
```
[the remark this answers is not in the document]
Comment (fcfargo) [reply products/urls.py]: 이 부분도 수정 반영이 안된 것 같습니다. 일단 다시 한 번 제 코드를 확인하겠습니다.

File: products/views.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ulfrid) [reply products/views.py]: 아 이건 제가 기존에 작성했는데 코드가 변경되서 컨트롤 안된게 남은거에요 ㅋㅋ

File: products/views.py (hunk 1 above)
Comment (fcfargo) [reply products/views.py]: 예외처리 반영하도록 하겠습니다.

File: products/views.py (hunk 3 above)
Comment (fcfargo) [reply products/views.py:85]: 바로 위에서 지적해주신 내용까지 포함하여 말씀드립니다. 일단 멘토님께서 지적해주신 것 처럼, 프로덕트 객체 자체의  id값을 전달하는 것이 바람직할 것 같습니다. 생각해보니 장바구니 담기 요청이 이뤄지기 위해선 반드시 필요한 데이터였습니다ㅠㅠ. 수정하도록 하겠습니다. 다만,  id 값을 전달하는 이유는 프론트엔드의 요구가 반영된 결과물입니다. 프론트엔드에서 백엔드로부터 받은 데이터를 처리하려면 순차적으로 부여된 id값이 따로 필요하다고 하더군요. 그래서 저렇게 id 값을 넣게 됐습니다.

File: products/views.py (hunk 4 above)
Comment (fcfargo) [reply products/views.py]: 수정 반영하겠습니다.   MtoM으로 모델링을 했음에도 불구하고, 실제 데이터를 가져올 때 활용하질 못했네요...

File: products/views.py (hunk 3 above)
Comment (devops-andrew) [reply products/views.py:85]: 넵 나중에 사이드 이펙트가 없으려면 정확한 구분이 좋습니다~!

**Review by devops-andrew — APPROVED** [review]
LGTM 👍

**Pull request merged**